### DISCUSSION TITLE
Split UI Features from Logic

### DISCUSSION THREAD

**Discussion by scriptsengineer** [Ark2000/PankuConsole#135] [Ideas]
### Problem
I've been testing the console for some time and I realize this big dilemma for its flexible use in final projects
- UI customization
- Dependence on the current UI architecture assembled

Separating the logic from the UI would solve this, one plan is for the UI to read values and signals from the PankuConsole singleton and subscribe to signals from the singleton responding if necessary by directly called methods.
The Core wouldn't need to know if there is a UI or not.

> Note: This has already been done in the console module, however the dependence on the use of the UI together with the singleton remained rigid.

The biggest benefit of this approach is that UI elements can be placed anywhere in the node tree, which could be within a game UI for example, or within post-processing. The core does not need to know whether or not a UI exists.

### Plan

I intend to start this work today mainly understanding how the modules work, for me this is the biggest dilemma that makes it difficult to easily transform the UI into a separate element in the architecture. Maybe each module will also add a UI element if necessary.

I'm open to ideas on how to solve it and I also want to question some elements of the addon in the future to make their importance clear so I can split this safely without breaking anything.

**Comment by Ark2000** [comment]
Speaking of UI in PankuConsole, here are some key points:

- The core itself only provides a multi-windows UI system
- Each module has full control over its own UI.
- The interdependence between modules is defined through the core's event bus and APIs, although direct module access is also possible through the module manager.
- The Data Controller module can be utilized to generate a data panel ui window. (e.g. modules can define `opt.gd` to describe module options)

I'm still not fully clear on the specific problem you're facing. Are you planning to develop a new module with a custom UI, or are you experiencing difficulties in customizing the data controller? Providing an example of the specific obstacle you're encountering would be helpful in understanding the issue better.

**Comment by scriptsengineer** [comment]
The difficulty is customizing the UI without having to change something in the project, or keeping the UI in a place separate from the singleton, for example I want the UI to be part of something in the game or to be inside a node that customizes the visualization with a shader

**Reply by Ark2000** [reply]
well. it seems out of my initial thought, the plugin's ui is completely designed around the multi-windows concept. which will be a little troubling in regard to your desire. You may have to customize the multi-window system heavily.

**Reply by Ark2000** [reply]
Also, have you ever played the game EVE Online, it has a fantastic UI design which is the main inspiration of the ui system, I wish to be closer toward which in the future development

**Reply by scriptsengineer** [reply]
Really, when I see the code I see that it is almost unfeasible without breaking something that already exists. even this reason to open the discussion and not an "issue"
In any case, I'm open to any kind of suggestion.
